experiment/cherrypicker/server.go

```diff
@@ -0,0 +1,399 @@
+	if !ok {
+		return
+	}
+	fmt.Fprint(w, "Event received. Have a nice day.")
+
+	if err := s.handleEvent(eventType, eventGUID, payload); err != nil {
+		logrus.WithError(err).Error("Error parsing event.")
+	}
+}
+
+func (s *Server) handleEvent(eventType, eventGUID string, payload []byte) error {
+	switch eventType {
+	case "issue_comment":
+		var ic github.IssueCommentEvent
+		if err := json.Unmarshal(payload, &ic); err != nil {
+			return err
+		}
+		go func() {
+			if err := s.handleIssueComment(ic); err != nil {
+				s.log.WithError(err).Info("Cherry-pick failed.")
+			}
+		}()
+	case "pull_request":
+		var pr github.PullRequestEvent
+		if err := json.Unmarshal(payload, &pr); err != nil {
+			return err
+		}
+		go func() {
+			if err := s.handlePullRequest(pr); err != nil {
+				s.log.WithError(err).Info("Cherry-pick failed.")
+			}
+		}()
+	default:
+		return fmt.Errorf("received an event of type %q but didn't ask for it", eventType)
+	}
+	return nil
+}
+
+func (s *Server) handleIssueComment(ic github.IssueCommentEvent) error {
+	// Only consider new comments in PRs.
+	if !ic.Issue.IsPullRequest() || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	cherryPickMatches := cherryPickRe.FindAllStringSubmatch(ic.Comment.Body, -1)
+	if len(cherryPickMatches) == 0 || len(cherryPickMatches[0]) != 2 {
+		return nil
+	}
+	targetBranch := cherryPickMatches[0][1]
+
+	if targetBranch == "master" {
+		return s.ghc.CreateComment(org, repo, num, "I don't cherrypick pull requests onto master.")
+	}
+
+	if ic.Issue.State != "closed" {
+		// Only members should be able to do cherry-picks.
+		ok, err := s.ghc.IsMember(org, commentAuthor)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			resp := fmt.Sprintf("Only [%s](https://github.com/orgs/%s/people) org members may request cherry picks. You can still do the cherry-pick manually.", org, org)
+			s.log.Info(resp)
+			return s.ghc.CreateComment(org, repo, num, resp)
+		}
+		resp := fmt.Sprintf("@%s once the present PR merges, I will cherry-pick it on top of %s in a new PR and assign it to you.", commentAuthor, targetBranch)
+		return s.ghc.CreateComment(org, repo, num, resp)
+	}
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		return err
+	}
+	baseBranch := pr.Base.Ref
+
+	// Cherry-pick only merged PRs.
+	if !pr.Merged {
+		resp := "cannot cherry-pick an unmerged PR"
+		s.log.Info(resp)
+		return s.ghc.CreateComment(org, repo, num, resp)
+	}
+
+	if baseBranch != "master" {
+		resp := "The base branch needs to be master in order to cherry-pick to a different branch."
+		s.log.Info(resp)
+		return s.ghc.CreateComment(org, repo, num, resp)
+	}
+
+	return s.handle(org, repo, commentAuthor, baseBranch, targetBranch, num)
+}
+
+func (s *Server) handlePullRequest(pre github.PullRequestEvent) error {
```

Comment:
Reviewed this flow, looks good as a first pass. Would be nice to dedupe and filter in the future.

/lgtm